modules/tensor_mechanics/include/materials/ComputeThermalExpansionEigenstrainBase.h

```diff
@@ -26,38 +31,55 @@ class ComputeThermalExpansionEigenstrainBaseTempl
 
   ComputeThermalExpansionEigenstrainBaseTempl(const InputParameters & parameters);
 
-protected:
+  /// resize _temperature_buffer
+  virtual void subdomainSetup() final;
+
+  /// update _temperature_buffer
+  virtual void computeProperties() final;
+
   virtual void computeQpEigenstrain() override;
-  /*
-   * Compute the total thermal strain relative to the stress-free temperature at
-   * the current temperature, as well as the current instantaneous thermal
-   * expansion coefficient.
-   * param thermal_strain    The current total linear thermal strain
-   *                         (\delta L / L)
-   * param dthermal_strain_dT The current instantaneous coefficient of thermal
-   *                         expansion (derivative of thermal_strain wrt
-   *                         temperature
+
+protected:
+  /**
+   * computeThermalStrain must be overridden in derived classes. The return type
+   * ValueAndDerivative<is_ad> contains the value for the thermal strain and its
+   * temperature derivative. Derived classes should use `_temperature[_qp]` to obtain
+   * the current temperature. In the is_ad == false case that member variable is
+   * agumented and will be of the type ChainedReal. I.e. even with is_ad == false
+   * a variant of forward mode automatic differentiation will be used internally to
+   * compute the thermal strain and no manual implementation of the temperature derivative
+   * is needed.
    */
-  virtual void computeThermalStrain(GenericReal<is_ad> & thermal_strain,
-                                    Real * dthermal_strain_dT = nullptr) = 0;
+  virtual ValueAndDerivative<is_ad> computeThermalStrain() = 0;
+
+  /**
+   * Temperature to use in the eigenstrain calculation (depending on _use_old_temperature).
```

Comment:
I'd make this more explicitly clear:
Temperature used in the eigenstrain calculation (current value if _use_old_temperature=false, old value if _use_old_temperature=true).